src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (key in currentState) {
+            const { [key]: _, ...rest } = currentState;
+
+            currentState = rest;
+          }
+        }
+        break;
+    }
```

Comment:
Consider adding a default case in the switch statement for error handling. It's a good practice to have a default case to handle unexpected action types.